compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -819,7 +819,27 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         }
 
         match self.candidate_from_obligation(stack) {
-            Ok(Some(c)) => self.evaluate_candidate(stack, &c),
+            Ok(Some(c)) => {
+                let mut res = self.evaluate_candidate(stack, &c)?;
+                // If we erased any lifetimes, then we want to use
+                // `EvaluatedToOkModuloRegions` instead of `EvaluatedToOk`
+                // as your final result. The result will be cached using
+                // the freshened trait predicate as a key, so we need
+                // our result to be correct by *any* choice of original lifetimes,
+                // not just the lifetime choice for this particular (non-erased)
+                // predicate.
+                // See issue #80691
+                let unbound_input_lifetimes = stack
+                    .fresh_trait_ref
+                    .skip_binder()
+                    .substs
+                    .regions()
+                    .any(|r| matches!(r, ty::RegionKind::ReErased));
```

Comment:
Can't types or consts also have erased regions nested in them? Ideally I feel like this should check a type flag - but @nikomatsakis / @matthewjasper would know better.

Reply:
That seems right - we would want to catch something like `T as MyTrait<MyType<'erased>>`